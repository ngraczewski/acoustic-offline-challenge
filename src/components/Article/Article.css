@import '../../css/variables.css';

.article {
  border: var(--border);
  border-radius: var(--border-radius);
  box-shadow: var(--box-shadow);
}

.article__lead-image {
  position: relative;
  border-top-left-radius: var(--border-radius);
  border-top-right-radius: var(--border-radius);
  width: 100%;
}

.article__header {
  display: flex;
  align-items: baseline;
  padding: 0 var(--spacing);
}

.article__header__date,
.article__header__author {
  color: var(--secondary-color);
}

.article__header__author {
  margin-left: auto;
}

.article__body {
  padding: 0 var(--spacing);
}
