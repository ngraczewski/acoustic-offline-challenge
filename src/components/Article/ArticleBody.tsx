import React from 'react';

export type Props = {
  body: string[];
};

export const ArticleBody = ({ body }: Props): JSX.Element => {
  return (
    <div className="article__body">
      {body.map((b, index) => (
        // since we are not actively modifying this list, using index as a key
        // is sufficent, in other cases something more coupled with the item
        // would be required
        <div key={index} dangerouslySetInnerHTML={{ __html: b }} />
      ))}
    </div>
  );
};
