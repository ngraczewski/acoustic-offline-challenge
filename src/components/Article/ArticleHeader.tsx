import React from 'react';

export type Props = {
  title: string;
  author: string;
  date: Date;
};

export const ArticleHeader = ({ title, author, date }: Props): JSX.Element => {
  const formattedDate = new Date(date).toDateString();
  return (
    <header className="article__header">
      <h1 className="article__header__title">{title}</h1>
      <div className="article__header__author">
        by {author} at {formattedDate}
      </div>
    </header>
  );
};
