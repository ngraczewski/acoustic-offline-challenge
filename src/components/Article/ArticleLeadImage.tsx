import React from 'react';

export type Props = {
  imageSrc: string;
  imageAltText: string;
};

export const ArticleLeadImage = ({
  imageAltText,
  imageSrc,
}: Props): JSX.Element => {
  return (
    <img
      className="article__lead-image"
      alt={imageAltText}
      src={imageSrc}
    ></img>
  );
};
