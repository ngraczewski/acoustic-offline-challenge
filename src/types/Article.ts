export type ArticleContent = {
  title: string;
  author: string;
  date: Date;
  body: string[];
  imageSrc: string;
  imageAltText: string;
};
