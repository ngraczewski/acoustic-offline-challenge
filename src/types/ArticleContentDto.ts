import { TextElement } from './TextElement';
import { FormattedTextElement } from './FormattedTextElement';
import { DateTimeElement } from './DateTimeElement';
import { ContentDto } from './ContentDto';
import { GroupElement } from './GroupElement';
import { ImageElement } from './ImageElement';
import { Rendition } from './Rendition';

export type ArticleContentDto = ContentDto<
  'Article',
  {
    heading: TextElement;
    author: TextElement;
    body: FormattedTextElement;
    date: DateTimeElement;
    mainImage: GroupElement<{
      leadImage: ImageElement<{
        lead: Rendition;
        card: Rendition;
        default: Rendition;
      }>;
      leadImageCaption: TextElement;
      leadImageCredit: TextElement;
    }>;
  }
>;
