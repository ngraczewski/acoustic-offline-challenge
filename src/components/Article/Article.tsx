import React from 'react';
import { ArticleHeader } from './ArticleHeader';
import { ArticleLeadImage } from './ArticleLeadImage';
import { ArticleContent } from '../../types/Article';
import { ArticleBody } from './ArticleBody';

import './Article.css';

export const Article = ({
  title,
  author,
  body,
  date,
  imageAltText,
  imageSrc,
}: ArticleContent): JSX.Element => {
  return (
    <div className="article">
      <ArticleLeadImage imageSrc={imageSrc} imageAltText={imageAltText} />
      <ArticleHeader title={title} author={author} date={date} />
      <ArticleBody body={body} />
    </div>
  );
};
