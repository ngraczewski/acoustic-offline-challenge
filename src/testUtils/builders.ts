import faker from 'faker';
import { merge } from 'lodash';
import { ContentDto } from '../types/ContentDto';
import { ArticleContent } from '../types/Article';
import { ArticleContentDto } from '../types/ArticleContentDto';
import { Rendition } from '../types/Rendition';
import { DeepPartial } from '../types/DeepPartial';

// https://github.com/Marak/faker.js/issues/608
export const buildDate = () => faker.date.past(1, new Date(2020, 0, 0));

export const buildContentDto = (
  override?: Partial<ContentDto>
): ContentDto => ({
  classification: faker.random.words(),
  created: faker.date.recent().toISOString(),
  creatorId: faker.random.uuid(),
  description: faker.random.words(),
  id: faker.random.words(),
  kind: [faker.random.words(), faker.random.words()],
  lastModified: faker.date.recent().toISOString(),
  lastModifierId: faker.random.uuid(),
  locale: faker.random.locale(),
  name: faker.random.words(),
  tags: [faker.random.word(), faker.random.word()],
  typeId: faker.random.uuid(),
  type: faker.random.word(),
  elements: {},
  ...override,
});

export const buildArticleContent = (
  override?: Partial<ArticleContent>
): ArticleContent => ({
  title: faker.lorem.sentence(),
  author: faker.name.firstName(),
  date: buildDate(),
  imageAltText: faker.lorem.sentence(),
  imageSrc: faker.internet.url(),
  body: [
    `<p>${faker.lorem.paragraph()}</p>`,
    `<p>${faker.lorem.paragraph()}</p>`,
  ],
  ...override,
});

export const buildRendition = (override?: Partial<Rendition>): Rendition => ({
  url: faker.internet.url(),
  height: faker.random.number(),
  width: faker.random.number(),
  source: faker.internet.url(),
  ...override,
});

export const buildArticleContentDto = (
  override?: DeepPartial<ArticleContentDto>
): ArticleContentDto =>
  merge(
    {
      ...buildContentDto(),
      type: 'Article',
      elements: {
        author: {
          elementType: 'text',
          value: faker.name.firstName(),
        },
        heading: {
          elementType: 'text',
          value: faker.lorem.sentence(),
        },
        body: {
          elementType: 'formattedtext',
          values: [
            `<p>${faker.lorem.paragraph()}</p>`,
            `<p>${faker.lorem.paragraph()}</p>`,
          ],
        },
        date: {
          elementType: 'datetime',
          value: buildDate().toISOString(),
        },
        mainImage: {
          elementType: 'group',
          value: {
            leadImageCaption: {
              elementType: 'text',
              value: faker.random.word(),
            },
            leadImageCredit: {
              elementType: 'text',
              value: faker.random.word(),
            },
            leadImage: {
              asset: {
                altText: faker.lorem.sentence(),
                fileName: faker.system.fileName(),
                height: faker.random.number(),
                width: faker.random.number(),
                id: faker.random.uuid(),
                mediaType: faker.system.mimeType(),
                resourceUri: faker.internet.url(),
              },
              url: faker.internet.url(),
              altText: faker.random.word(),
              elementType: 'image',
              mode: 'shared',
              profiles: [faker.random.uuid(), faker.random.uuid()],
              renditions: {
                card: buildRendition(),
                default: buildRendition(),
                lead: buildRendition(),
              },
            },
          },
        },
      },
    },
    override
  );
