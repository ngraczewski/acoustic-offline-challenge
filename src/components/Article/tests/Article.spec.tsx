import React from 'react';
import faker from 'faker';
import { shallow } from 'enzyme';
import { Article } from '../Article';
import { ArticleContent } from '../../../types/Article';
import { buildArticleContent } from '../../../testUtils/builders';

describe('Article', () => {
  beforeEach(() => {
    faker.seed(1);
  });

  it('should render correctly', () => {
    const props: ArticleContent = buildArticleContent();

    const wrapper = shallow(<Article {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
