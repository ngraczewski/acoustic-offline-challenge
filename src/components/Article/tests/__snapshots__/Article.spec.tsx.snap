// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article should render correctly 1`] = `
<div
  className="article"
>
  <ArticleLeadImage
    imageAltText="Sunt qui amet iure."
    imageSrc="http://marcellus.info"
  />
  <ArticleHeader
    author="Zola"
    date={2019-11-07T09:25:05.320Z}
    title="Repellat quisquam recusandae alias consequuntur corporis."
  />
  <ArticleBody
    body={
      Array [
        "<p>Nulla quam ipsam nobis cupiditate sed dignissimos. Incidunt accusantium sed libero repudiandae esse blanditiis natus et eos. Velit omnis et porro ut et ipsam explicabo eligendi occaecati. Et saepe eum dicta eum eaque enim ipsum inventore debitis. Aspernatur deserunt quam tempore a velit provident velit. Nostrum ipsam qui nobis repellendus fugiat velit sit.</p>",
        "<p>Fuga doloribus distinctio placeat ullam minima ducimus temporibus modi. Architecto ducimus totam. Explicabo exercitationem ut quis sed magni cupiditate sit vitae soluta.</p>",
      ]
    }
  />
</div>
`;
